src/use-swr.ts

```diff
@@ -505,15 +506,27 @@ function useSWR<Data = any, Error = any>(
         }
       }
 
-      if (timeout !== null) {
-        clearTimeout(timeout)
-      }
-
       if (reconnect !== null) {
         removeEventListener('online', reconnect)
       }
     }
-  }, [key, config.refreshInterval, revalidate])
+  }, [key, revalidate])
+
+  // set up polling
+  const refresh = async () => {
+    if (
+      !errorRef.current &&
+      (config.refreshWhenHidden || isDocumentVisible()) &&
+      (!config.refreshWhenOffline && isOnline())
+    ) {
+      // only revalidate when the page is visible
+      // if API request errored, we stop polling in this round
+      // and let the error retry function handle it
+      await revalidate({ dedupe: true })
+    }
+  }
```

Comment:
`refresh` will be recreated in every render (and triggers the `useIsomorphicLayoutEffect` inside `usePolling`) so might not be efficient.

A suggestion is to add deps of `refresh` to `usePolling`:
```js
const refresh = ...

usePolling(refresh, config.refreshInterval, [config.refreshWhenHidden, config.refreshWhenOffline, revalidate]))
```

And inside `usePolling`, we can use the deps:

```js
  useIsomorphicLayoutEffect(() => {
    callbackRef.current = callback
  }, deps)
```

This way there're only 2 `useEffect`s being used.